### PR TITLE
fix(qmmm): make embedding=mechanical runnable (zero embedding field)

### PR DESCRIPTION
**Pull request by karmachoi**
## The bug

`[qmmm] embedding = mechanical` aborts on the first SCF iteration:

```
qmmm_espf::add_potqm_contributions: Record `OQP::POTQM` not found!
```

`scf.F90:1207` calls `add_potqm_contributions` on **every** SCF iteration whenever `control%qmmm_flag` is set, and that routine requires `OQP::SM`, `OQP::ESPF_CORR` and `OQP::POTQM` with `WITH_ABORT`. But `compute_force` only builds the embedding arrays for the electrostatic/split/ESPF paths:

```python
potmm = potqm = None
if self.Embedding in ("electrostatic", "split") or self.espf_full:
    potmm, potqm = self.electrostatic_potential()
```

so mechanical leaves `potmm = potqm = None`, the `OQP::POTQM` record is never created, and the SCF dies. The unconditional `oqp.grad_esp_qmmm()` call needs `OQP::POTMM` for the same reason.

This is the same family as the known single-point QM/MM gap (`runtype=energy` + `qmmm_flag=true` aborting on a missing `OQP::ESPF_CORR`): `scf.F90` assumes `qmmm_flag` implies the ESPF records exist, which only the ESPF path guarantees.

## The fix

Give mechanical embedding a **zero field** rather than an absent one. That is precisely what mechanical embedding means for the QM subsystem: the QM Hamiltonian is unperturbed by the MM charges (a gas-phase SCF), and QM–MM electrostatics is left to OpenMM via the QM ESP charges in `forces_mm()`.

A zero field is exact, not an approximation — it adds 0 to `hcore`, contributes 0 to the ESPF gradient, and leaves `eqm` untouched, so the existing `if potmm is not None` guards remain numerically no-ops.

Note `potqm` is already zeroed unconditionally in `forces_qm_openqp` (the #205 M1a fix), so `add_potqm_contributions` is a no-op in **every** code path today — the record only has to exist. A Fortran-side early return on absent records would be an equally valid fix and would avoid the (small) wasted work of building the ESPF grid; happy to switch if maintainers prefer that. `espf_op_corr` is needed regardless, since `form_esp_charges` supplies the ESP charges the mechanical path hands to OpenMM.

## Verification

Test system: methanol with QM = OH, MM = CH3. The partition cuts the C–O bond, so one hydrogen link atom caps the O (`g = 0.7152`). HF/6-31G, NVE + Verlet, dt = 0.25 fs.

- **Before:** aborts on the first SCF iteration.
- **After:** completes 400 NVE steps.

## Scope caveat — this makes mechanical *run*, not *exact*

`forces_mm()` writes geometry-dependent QM ESP charges into OpenMM, whose force is evaluated at **fixed** charges, so the charge-response term `Σ_A (∂E_mm/∂q_A)(dq_A/dR)` is never computed. Finite differences on the methanol system:

| quantity | O_x | HO_x | C_x | H1_x |
|---|---|---|---|---|
| `-gmm` (OpenMM) | 91.78 | 11.06 | −88.48 | 161.50 |
| FD, frozen charges | 91.76 | 11.06 | −88.46 | 161.50 |
| FD, charges follow geometry | **−243.97** | **282.86** | **−24.51** | 161.50 |

OpenMM reproduces the frozen-charge derivative to 0.02%, but the true derivative differs by up to ~336 kJ/mol/nm. The methyl hydrogens, which carry no QM-dependent charge, are exact — as expected.

This is a pre-existing property of the split/mechanical design, independent of link atoms, and is **left untouched here**. Mechanical embedding should not be treated as gradient-exact for dynamics on the strength of this PR; the ESPF path avoids the issue by zeroing the QM charges in OpenMM and handling charge fluctuation through `grad_esp_qmmm`.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: pyoqp/oqp/library/qmmm_driver.py [hunk 1]
```diff
@@ -537,6 +537,18 @@ def compute_force(self, positions, topology, mm_systems, qm_atoms):
...
+            # SCF is gas-phase and QM-MM electrostatics is left to OpenMM (via
+            # the QM ESP charges in forces_mm). The embedding arrays must still
+            # exist and be ZERO rather than absent: scf.F90 calls
+            # add_potqm_contributions on every SCF iteration whenever
+            # qmmm_flag is set, and that routine aborts on a missing
+            # OQP::POTQM record ("Record `OQP::POTQM` not found!"). The same
+            # applies to OQP::POTMM via the unconditional grad_esp_qmmm call.
+            # A zero field reproduces gas-phase QM exactly: it adds 0 to hcore,
+            # contributes 0 to the ESPF gradient, and leaves eqm untouched.
+            potmm, potqm = self._zero_embedding()
```
[the remark this answers is not in the document]
Comment (karmachoi) [reply pyoqp/oqp/library/qmmm_driver.py:574]: Fixed in d929d0425. Native AO methods still receive explicit zero POTMM/POTQM records for mechanical QM/MM, while the DFTB dispatch translates that mechanical mode back to potmm=None, preserving the DFTB gas-phase contract. The targeted dispatch regression passes.

File: pyoqp/oqp/library/qmmm_driver.py [hunk 2]
```diff
@@ -537,6 +537,18 @@ def compute_force(self, positions, topology, mm_systems, qm_atoms):
         potmm = potqm = None
         if self.Embedding in ("electrostatic", "split") or self.espf_full:
             potmm, potqm = self.electrostatic_potential()
+        else:
```
[the remark this answers is not in the document]
Comment (karmachoi) [reply pyoqp/oqp/library/qmmm_driver.py]: Fixed in d929d0425. Embedding values are normalized once and validated against mechanical/electrostatic/espf/espf_full/split; unknown spellings now raise ValueError instead of falling through to mechanical. Regressions cover rejection and case normalization.

**Comment by mohsenkor** [comment]
Reviewed by Claude and Codex independently. **CI is green now** — the 4 failures were not caused by this PR.

## The red CI was a stale base, not this change

The failing jobs were the example regression suite: `C2H4_BHHLYP-MRSFTDDFT_MECI`, `_MECP_OQP`, `_TCI_OQP`, `_BAEKA_OQP` and `H2O_RHF-DFT_OPTIMIZE`, all failing on coord/energy/grad tolerances. This PR touches two files — `qmmm_driver.py` and `test_qmmm_dftb.py` — and neither is reachable from those examples.

The branch was **48 commits behind `main`**, based on `e5f774f` (#272). The commits since include the TRAH recovery work (#287, #288), which is exactly what those optimization/MECI convergence failures needed. Merged current `main` in; all four now pass.

## The zero-field argument holds

Codex checked the Fortran rather than taking the PR description's word for it:

- `add_potqm_contributions` consumes POTQM only via `matmul(potqm, q)` plus a linear `dhcore` correction (`qmmm.F90:335`), and its `dhcore` buffer is dedicated and zero-initialised.
- POTMM enters hcore, the energy, and the ESPF gradient linearly.
- `form_esp_charges` consumes neither POT record (`qmmm.F90:583`).

So zero arrays leave energies and gradients unchanged: **"exact, not an approximation" is correct**. The ESPF machinery still runs, but contributes exactly zero.

Independently checked on the Python side:
- `_zero_embedding` shapes `(nqm+nlink,)` / `(nqm+nlink, nqm+nlink)` match both `electrostatic_potential`'s full-field return and `_pad_potential_for_link_atoms`; dtype is float64 as the einsum needs.
- Every `if potmm is not None` guard now executes but adds exactly zero.
- The in-place `potmm -=` mutation is safe — `compute_force` doesn't read `potmm` after the call.
- `_normalize_embedding` breaks nothing: the schema default is `electrostatic` (`oqpdata.py:109`), and every config, example, and call site in the repo uses an accepted value. Arbitrary strings previously fell through silently as mechanical, which was undocumented typo tolerance.
- DFTB dispatch passing `None` is right for both mol and config modes — `_forces_qm_dftb` maps it to `mol.dftb_external_potential = None`.

## Test gaps fixed (575f3d4)

Codex flagged the tests as mock-only. A related problem: **the two `_normalize_embedding` tests never ran anywhere.** They sat in a class gated on `_HAVE_OPENMM and _HAVE_OQP`, and CI installs the compiled backend but not OpenMM — so a string-validation test with no simulation dependency was skipped in CI as well as locally.

- Moved them to a class gated only on the backend, plus a case asserting every spelling the dispatch branches on is accepted.
- Added the electrostatic counterpart to the dispatch test — the mechanical-only test would also pass against a dispatch that hardcoded `None`.
- Added the missing `_zero_embedding` test: the shape contract keeps the `espf_op_corr` einsum aligned, and "the field is zero" is the entire correctness argument, so both are now asserted.

Still open from Codex, and worth doing before this is relied on: no test executes the actual `compute_force` mechanical branch against a real SCF. A native mechanical run compared to a gas-phase calculation, with and without a link atom, would pin the regression this PR fixes.

## Scope caveat — flagging for a maintainer decision

The PR is candid that this makes mechanical embedding *run*, not *exact*: OpenMM evaluates forces at frozen QM charges, so `Σ_A (∂E_mm/∂q_A)(dq_A/dR)` is never computed, with the author's own finite differences showing up to **~336 kJ/mol/nm** of error. That is pre-existing and untouched here, but it means mechanical embedding should not be used for dynamics on the strength of this PR. Whether that ships with the caveat or is addressed first is a scientific call, not a code-review one.
